src/member-api.js

```diff
@@ -54,42 +77,140 @@ export class MemberApi extends TypedEmitter {
    * @param {string} [opts.roleName]
    * @param {string} [opts.roleDescription]
    * @param {number} [opts.timeout]
-   *
-   * @returns {Promise<import('./generated/rpc.js').InviteResponse_Decision>}
+   * @returns {Promise<(
+   *   typeof InviteResponse_Decision.ACCEPT |
+   *   typeof InviteResponse_Decision.REJECT |
+   *   typeof InviteResponse_Decision.ALREADY
+   * )>}
    */
-  async invite(deviceId, { roleId, roleName, roleDescription, timeout }) {
+  async invite(
+    deviceId,
+    {
+      roleId,
+      roleName = ROLES[roleId]?.name,
+      roleDescription,
+      timeout = DEFAULT_INVITE_TIMEOUT,
+    }
+  ) {
     assert(isRoleIdForNewInvite(roleId), 'Invalid role ID for new invite')
+    assert(
+      !this.#deviceIdsWithPendingInvites.has(deviceId),
+      'Already inviting this device ID'
+    )
+    assert(timeout > 0, 'Timeout should be a positive number')
+    assert(timeout < 2 ** 32, 'Timeout is too large')
 
-    const { name: deviceName } = await this.getById(this.#ownDeviceId)
+    /**
+     * These cleanup functions should be synchronous so we don't have to
+     * `await` calls to `handleAborted`.
+     * @type {Array<() => void>}
+     */
+    const cleanupFns = []
+    const cleanup = () => {
+      for (const fn of cleanupFns) {
+        try {
+          fn()
+        } catch (_err) {
+          console.error('Cleanup function failed')
+        }
+      }
+    }
 
-    // since we are always getting #ownDeviceId,
-    // this should never throw (see comment on getById), but it pleases ts
-    if (!deviceName)
-      throw new Error(
-        'Internal error trying to read own device name for this invite'
-      )
+    try {
+      let handleAborted = noop
+      const timeoutId = setTimeout(() => {
+        handleAborted = () => {
```

Comment:
I think you need to abort the abortSignal in this timeout?

Reply:
That's done here:

https://github.com/digidem/mapeo-core-next/blob/210ab8bca6ed1f415daf144396b64f8354a8d56d/src/member-api.js#L155-L157

Reply:
yeah but if we never receive an invite response, `await receiveInviteResponsePromise` will never resolve, and the cleanup functions will never be called. Or am I missing something?

Reply:
I _think_ that sequence of events would be:

1. We set up the receive invite abort controller.
   https://github.com/digidem/mapeo-core-next/blob/25d3c073cb9992ea66f0ecf381bcf5efb2819cea/src/member-api.js#L154
1. We immediately add it to our list of cleanup functions.
   https://github.com/digidem/mapeo-core-next/blob/25d3c073cb9992ea66f0ecf381bcf5efb2819cea/src/member-api.js#L155-L157
1. The timeout fires and `handleAborted` now throws.
   https://github.com/digidem/mapeo-core-next/blob/25d3c073cb9992ea66f0ecf381bcf5efb2819cea/src/member-api.js#L121-L125
1. `handleAborted` is called, which throws an error.
   https://github.com/digidem/mapeo-core-next/blob/25d3c073cb9992ea66f0ecf381bcf5efb2819cea/src/member-api.js#L179
1. The error is caught and cleaned up.
   https://github.com/digidem/mapeo-core-next/blob/25d3c073cb9992ea66f0ecf381bcf5efb2819cea/src/member-api.js#L211-L213

More broadly: even if this code is technically correct[^0], it's hard to follow. I spent a good amount of time trying to make this clearer but I know I missed at least one good idea: `onceSatisfied`. I'll try to think of further ways to simplify this, as it's hard to understand.

[^0]: which it may not be

Reply:
The code that is going to 
potentially take a long time and we want to timeout is awaiting the invite to be accepted on line 181. If that never resolves, then I don't see how the timeout can abort it, since the abort only happens if the code keeps executing.

Reply:
You're totally right. Fixed in dbfb6a915912aca53dde9a0a3225c511d5408ac5.